File: python/quicksort.py
import random
def partition(arr, low, high):
    pivot_index = random.randint(low, high)
    pivot = arr[pivot_index] # Choosing pivot randomly
    swap(arr, pivot_index, high)
    i = low
    j = high
    while i < j:
        while arr[i] <= pivot and i < j:
            i += 1
        while arr[j] >= pivot and i < j:
            j -= 1
        swap(arr, i, j)
    swap(arr, i, high)
    return i


def quick_sort(arr, low, high):
    if low >= high:
        return
    index = partition(arr, low, high)
    quick_sort(arr, low, index-1)
    quick_sort(arr, index +1, high)



def swap(arr, i, j):
    temp = arr[i]
    arr[i] = arr[j]
    arr[j] = temp




numbers = [2, 1, 8, 5, 4, 9, 12, 99, 34, 56, 74]
quick_sort(numbers, 0, len(numbers)-1)

print(numbers)

File: Java code/Leetcode-Examples/src/QuickSort.java
public class QuickSort {
    public void quicksort(int[] arr){
        quicksort(arr, 0, arr.length -1);
    }
    private void quicksort(int[] arr, int low, int high){
        if (low >= high){
            return;
        }
        int pivotIndex = partition(arr, low, high);
        quicksort(arr, low, pivotIndex-1);
        quicksort(arr, pivotIndex+1, high);

    }
    private int partition(int[] arr, int low, int high){
        int pivot = arr[high];
        int leftPointer = low;
        int rightPointer = high;
        while(leftPointer < rightPointer){
            while(arr[leftPointer] <= pivot && leftPointer < rightPointer){
                leftPointer++;
            }
            while(arr[rightPointer] >= pivot && leftPointer < rightPointer){
                rightPointer--;
            }
            swap(arr, leftPointer, rightPointer);
        }
        swap(arr, leftPointer, high);
        return leftPointer;
    }
    private void swap(int[] arr, int i, int j){
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
}
